### PR TITLE
[Driver][FPGA][SYCL] Add specific timing diagnostic for FPGA AOT

### PR DESCRIPTION
**Pull request by mdtoguchi**
There are cases in which the generation of FPGA device code from 'aoc'
is known to have timing issues.  Add a specific diagnostic from the
driver when this case is encountered.  A special return code (42) is
used to inform the driver of this situation.

Additionally in this situation, we also want to allow for the compilation
to continue so the user can use the generated binary knowing the timing
issue as stated.

### CODE REVIEW

**Comment by mdtoguchi** [comment]
Only `aoc` knows when this situation occurs - any suggestion how to handle testing on our side?  I will ask the PSG folks if there is a specific code sequence we can use to reproduce reliably.

**Review by AGindinson — COMMENTED** [review]
The approach LGTM overall, just a couple stylistic comments

**Review by AlexeySachkov — APPROVED** [review]
`llvm-foreach` part LGTM. However, it would be great to have a test for error code preserving mechanism

File: clang/include/clang/Driver/Job.h [hunk 1]
```diff
@@ -105,7 +105,7 @@ struct ResponseFileSupport {
 /// execute.
 class Command {
 public:
-  using ErrorCodeDiagMapTy = llvm::DenseMap<int, std::string>;
+  using ErrorCodeDiagMapTy = llvm::DenseMap<int, std::pair<std::string, bool>>;
```
[the remark this answers is not in the document]
Comment (mdtoguchi) [reply clang/include/clang/Driver/Job.h]: Done.

File: clang/include/clang/Driver/Job.h [hunk 2]
```diff
@@ -194,15 +196,20 @@ class Command {
   virtual int Execute(ArrayRef<Optional<StringRef>> Redirects,
                       std::string *ErrMsg, bool *ExecutionFailed) const;
 
-  /// Store a custom driver diagnostic message upon a particular error code
-  /// returned by the command
-  void addDiagForErrorCode(int ErrorCode, StringRef CustomDiag);
+  /// Store a custom driver diagnostic message and if the compilation should
+  /// exit upon a particular error code returned by the command
+  void addDiagForErrorCode(int ErrorCode, StringRef CustomDiag,
+                           bool NoExit = false);
```
[the remark this answers is not in the document]
Comment (mdtoguchi) [reply clang/include/clang/Driver/Job.h]: Done.

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 3]
```diff
@@ -119,8 +130,14 @@ void SYCL::constructLLVMForeachCommand(Compilation &C, const JobAction &JA,
...
   llvm::sys::path::append(ForeachPath, "llvm-foreach");
   const char *Foreach = C.getArgs().MakeArgString(ForeachPath);
-  C.addCommand(std::make_unique<Command>(JA, *T, ResponseFileSupport::None(),
-                                         Foreach, ForeachArgs, None));
+
+  auto Cmd = std::make_unique<Command>(JA, *T, ResponseFileSupport::None(),
+                                       Foreach, ForeachArgs, None);
+  // FIXME: Add the FPGA specific timing diagnostic to the foreach call.
+  // The foreach call obscures the return codes from the tool it is calling
+  // to the compiler itself.
```
Comment (AGindinson) [inline clang/lib/Driver/ToolChains/SYCL.cpp:138]: To clarify, is it the behavior of `llvm-foreach` that should FIXed?

**Review by AGindinson — DISMISSED** [review]
LGTM.

File: clang/lib/Driver/ToolChains/SYCL.cpp (hunk 3 above)
Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/SYCL.cpp:138]: Right - I have already made the needed adjustments to pass back the error code from the tool being called through `llvm-foreach` but having to add the diagnostic code check to both the tool call and `llvm-foreach` is not efficient and can present issues if we have multiple tools using the same return code, and those tools go through `llvm-foreach`

**Comment by mdtoguchi** [comment]
@AlexeySachkov, can you take another look?

**Comment by AGindinson** [comment]
@bader, can this be merged without `llvm-foreach` owners' approval, given that @AlexeySachkov and @Fznamznon are unavailable and have granted an approval previously with a minor (now-addressed) comment?

File: clang/include/clang/Driver/Job.h [hunk 4]
```diff
@@ -194,15 +200,23 @@ class Command {
   virtual int Execute(ArrayRef<Optional<StringRef>> Redirects,
                       std::string *ErrMsg, bool *ExecutionFailed) const;
 
-  /// Store a custom driver diagnostic message upon a particular error code
-  /// returned by the command
+  /// Store a custom driver diagnostic message and if the compilation should
+  /// exit upon a particular error code returned by the command
   void addDiagForErrorCode(int ErrorCode, StringRef CustomDiag);
```
Comment (bader) [inline clang/include/clang/Driver/Job.h:205]: This method implementation has not changed. I think the comment should be left as-is.
```suggestion
  /// Store a custom driver diagnostic message upon a particular error code
  /// returned by the command
  void addDiagForErrorCode(int ErrorCode, StringRef CustomDiag);
```

**Review by bader — COMMENTED** [review]
`llvm-foreach` changes look good to me, but one comment update seems to be unintentional. Right?

**Review by AGindinson — APPROVED** [review]

**Pull request merged**